### PR TITLE
fix(Core/Player): Weapon skill gain

### PR DESCRIPTION
**Pull request by Kitzunu**
… PvP

* cherry-pick https://github.com/TrinityCore/TrinityCore/commit/e20e51a3855595817df89ac69494b9fd41e1acb3

* cherry-pick https://github.com/TrinityCore/TrinityCore/commit/0b616e4865f89b1d15ae20dccd83825cddb5c288

Co-Authored-By: Shauren <shauren.trinity@gmail.com>
Co-Authored-By: Max Drosdo.www <9049790+ipriver@users.noreply.github.com>

## Changes Proposed:
-  Fix ranged weapon skillups
-  Dont allow skillup in pvp
- Improve the if to case
- Update unarmed skill with Fist Weapon (https://www.wowhead.com/spell=15590/fist-weapons#comments:id=972512)

## Issues Addressed:

- Closes https://github.com/azerothcore/azerothcore-wotlk/issues/5651

## Tests Performed:

- Tested on Training Dummy
- Tested in Duel
- Tested in BattleGround

## How to Test the Changes:

1. Test if you can get skill on training dummy
2. Test if you can get skill in pvp
3. Check if unarmed is increased with fist weapon

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

File: src/server/game/Entities/Object/Object.h [hunk 1]
```diff
@@ -266,6 +266,7 @@ class Object
     // FG: some hacky helpers
     void ForceValuesUpdateAtIndex(uint32);
 
+    inline bool IsPlayer() const { return GetTypeId() == TYPEID_PLAYER; }
```
Comment (FrancescoBorzi) [inline src/server/game/Entities/Object/Object.h:269]: please use `[[nodiscard]]` when applicable

**Comment by Azcobu** [comment]
Tested no skillups from training dummies:
- bows :heavy_check_mark:
- wands :heavy_check_mark:
- guns :heavy_check_mark:
- thrown weapons :heavy_check_mark:

Tested no skillups in PVP/duels :heavy_check_mark:
Tested fist weapons give Unarmed skill - possibly working too well, as attacks seem to give 2 points not 1;
WoWScrnShot_052121_211858
However, as the character is attacking with 2 fists at once, perhaps this is expected hehaviour?

Otherwise, it all looks fine.

**Comment by Kitzunu** [comment]
Yes

**Review by FrancescoBorzi — APPROVED** [review]

**Comment by FrancescoBorzi** [comment]
@Kitzunu let's also add [To be merged] together with [Ready to be reviewed] after a PR has been successfully tested, so we don't forget to review it and merge.

good job you and @Azcobu !

**Comment by Kitzunu** [comment]
Part https://github.com/azerothcore/azerothcore-wotlk/pull/6015

Dont merge yet, I will put the SQL in this PR to sync unarmed and fist weapon skill for characters

**Comment by Kitzunu** [comment]
Now it can be merged =)

**Pull request merged**